File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int dp_sol(int y, int x, int n, vector<int>& dp) {
    // base case
    if (y == x)
        return 0;
    if (y < x)
        return 1e9;

    // dp
    int& ret = dp[y];
    if (ret != -1)
        return ret;

    ret = 1e9;
    // n 빼기
    if (y - n >= x)
        ret = min(ret, dp_sol(y - n, x, n, dp) + 1);

    // 2 나누기
    if (y % 2 == 0 && y / 2 >= x)
        ret = min(ret, dp_sol(y / 2, x, n, dp) + 1);

    // 3 나누기
    if (y % 3 == 0 && y / 3 >= x)
        ret = min(ret, dp_sol(y / 3, x, n, dp) + 1);

    return ret;
}

int solution(int x, int y, int n) {
    vector<int> dp(y + 1, -1);
    int answer = dp_sol(y, x, n, dp);

    if (answer == 1e9)
        answer = -1;

    return answer;
}

File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.java
import java.util.*;

class Solution {
    public int solution(int x, int y, int n) {
        int answer = 0;
        int[] dp = new int[y + 1];
        Arrays.fill(dp, Integer.MAX_VALUE);
        
        dp[x] = 0;
        for(int i = x ; i <= y ; i++) {
            
            // n 더하기
            if(i - n >= 0 && dp[i - n] != Integer.MAX_VALUE){
                dp[i] = Math.min(dp[i], dp[i - n] + 1);
            }
            
            // 2 곱하기
            if(i / 2 >= 0 && i % 2 == 0 && dp[i / 2] != Integer.MAX_VALUE){
                dp[i] = Math.min(dp[i], dp[i / 2] + 1);
            }
            
            if(i / 3 >= 0 && i % 3 == 0 && dp[i / 3] != Integer.MAX_VALUE){
                dp[i] = Math.min(dp[i], dp[i / 3] + 1);
            }
            
        }
        
        return dp[y] == Integer.MAX_VALUE ? -1 : dp[y];
    }
}